### PR TITLE
Mobile, Desktop, Cli: Resolves #15714: Index locked note resources

### LINKED ISSUE #15714: Index locked note resources from extracted IDs
#15556 added `extracted_resource_ids` and #15597 fills it from the plaintext note body before the body is encrypted.

This issue is for using those saved IDs in `ResourceService`. A locked note's stored body is ciphertext, so the normal scanner cannot find its resource links. That can clear its `note_resources` associations and make attachments used only by that note look unused.

For locked notes, `indexNoteResources` should rebuild the associations from `extracted_resource_ids` using existing validation path. This also needs to work when note lock is disabled, since locked notes may already exist or arrive through sync.

Normal note indexing should stay unchanged, and this does not require changing `NoteResource.orphanResources()`.

Relevant proposal sections: [3.3, 3.4, and the cross-device case in 3.8](https://docs.google.com/document/d/1YU3lY1ZmtFvMTKZDUyRmX4cB7ly8pzO8ECu2xacfqng/edit?usp=sharing).

The proposal still uses the older `is_locally_encrypted` name, which is now `is_locked`.

### PR DESCRIPTION
**Pull request by keshav0479**
## Summary

Resolves #15714.

Uses `extracted_resource_ids` to rebuild resource associations for locked notes.

A locked note's stored body is ciphertext, so the normal resource scanner cannot read its links. Without the associations, attachments used only by that note could be treated as unused.

This handling stays active when note lock is disabled because locked notes may already exist locally or arrive through sync.

## Changes

- includes `is_locked` and `extracted_resource_ids` in the ResourceService note query
- rebuilds locked-note associations from `extracted_resource_ids` using the existing validation path
- keeps normal unlocked-note indexing unchanged
- allows locked E2EE notes to be indexed without blocking later note changes
- adds focused tests for feature-disabled behavior, invalid IDs, association updates, and cross-client sync

This does not change `NoteResource.orphanResources()`.

## Testing

- `node .yarn/releases/yarn-4.16.0.cjs workspace @joplin/lib tsc`
- `node .yarn/releases/yarn-4.16.0.cjs workspace @joplin/lib test --runInBand services/ResourceService.test.js`
- `node .yarn/releases/yarn-4.16.0.cjs eslint packages/lib/services/ResourceService.ts packages/lib/services/ResourceService.test.ts`

## Manual testing

Did a manual desktop pass with two clean profiles syncing through a filesystem target.

Profile A created a notebook and a note with an image attachment, then synced. Profile B synced from the same target and received the note and attachment correctly. I edited the note on profile B, synced again, then synced profile A and confirmed the edit came back with the attachment still present.

No sync errors.

https://github.com/user-attachments/assets/7374f798-af75-4b64-89f4-9416e7cd2fdc

## AI Assistance Disclosure

I used AI tools while working on this PR, mainly for code review, checking the diff for scope issues, and sanity-checking the test plan. I went through the changes myself, kept the PR limited to resource association indexing, and understand the code being submitted.

### CODE REVIEW

File: packages/lib/services/ResourceService.ts [hunk 1]
```diff
@@ -77,16 +77,19 @@ export default class ResourceService extends BaseService {
 						const note = noteById(change.item_id);
 
 						if (note) {
-							if (note.encryption_applied) {
+							if (note.is_locked) {
+								const noteBody = note.extracted_resource_ids.split(',').map((id: string) => `[](:/${id})`).join(' ');
+								await this.setAssociatedResources(note.id, noteBody);
```
[the remark this answers is not in the document]
Comment (keshav0479) [reply packages/lib/services/ResourceService.ts:83]: extracted_resource_ids is NOT NULL with an empty-string default in the local DB and missing values from older sync items are normalized before they're saved, since this query reads directly from the notes table, it should always be a string here.

File: packages/lib/services/ResourceService.test.ts [hunk 2]
```diff
@@ -136,6 +137,94 @@ describe('services/ResourceService', () => {
...
+				body: 'encrypted body',
+				is_locked: 1,
+				extracted_resource_ids: '',
+			});
+			await service.indexNoteResources();
+
+			expect(await NoteResource.associatedNoteIds(resource1.id)).toEqual([]);
+		} finally {
+			Setting.setConstant('env', Env.Dev);
+		}
```
[the remark this answers is not in the document]
Comment (keshav0479) [reply packages/lib/services/ResourceService.test.ts]: the shared test setup initializes env to Env.Dev, so resetting it here restores test environment's default state.

File: packages/lib/services/ResourceService.ts (hunk 1 above)
Comment (personalizedrefrigerator) [inline packages/lib/services/ResourceService.ts]: Rather than mapping IDs to note links, then parsing them back, it might be cleaner to directly call
```ts
await NoteResource.setAssociatedResources(noteId, resourceIds);
```

Comment (keshav0479) [reply packages/lib/services/ResourceService.ts]: yeah, that would be cleaner. one thing i wanted to check: `extracted_resource_ids` comes from `linkedItemIds`, so it can include note or folder IDs and `NoteResource.setAssociatedResources` doesn’t validate them, current round-trip filters those out through `linkedResourceIds`.

did you mean to trust the saved IDs here, or filter them against resources table first and then call `NoteResource.setAssociatedResources` directly?

Comment (personalizedrefrigerator) [reply packages/lib/services/ResourceService.ts]: > did you mean to trust the saved IDs here, or filter them against resources table first and then call NoteResource.setAssociatedResources directly?

Good point! I missed that `setAssociatedResources` does database queries (via `linkedResourceIds`) to check that the attached items are actually resources. In that case, it's probably best to continue using `this.setAssociatedResources`.

An alternative would be to call `loadItemsByIds` and filter the output by model type (as `linkedResourceIds` does), but this duplicates logic already present in `linkedResourceIds`/`linkedItemIdsByType`.

File: packages/lib/services/ResourceService.test.ts [hunk 3]
```diff
@@ -136,6 +137,94 @@ describe('services/ResourceService', () => {
...
+		const resource2 = await shim.createResourceFromPath(`${supportDir}/welcome.pdf`);
+		const note1 = await Note.save({
+			title: 'locked note',
+			parent_id: folder1.id,
+			body: `encrypted body [](:/${resource2.id})`,
+			is_locked: 1,
+			extracted_resource_ids: `${resource1.id},not-a-resource-id,${folder1.id}`,
+		});
+
+		Setting.setConstant('env', Env.Prod);
```
Comment (laurent22) [inline packages/lib/services/ResourceService.test.ts]: As mentioned in a previous PR, please create a feature flag so that it's easier to find what parts of your code are affected by it

File: packages/lib/services/ResourceService.ts [hunk 4]
```diff
@@ -77,16 +77,19 @@ export default class ResourceService extends BaseService {
 						const note = noteById(change.item_id);
 
 						if (note) {
-							if (note.encryption_applied) {
+							if (note.is_locked) {
+								const noteBody = note.extracted_resource_ids.split(',').map((id: string) => `[](:/${id})`).join(' ');
```
Comment (laurent22) [inline packages/lib/services/ResourceService.ts]: This `note.extracted_resource_ids.split(',')` would give `['']` if extracted_resource_ids is empty so I think you'll end up with empty resources in the note body. Please check it's not empty before proceeding.

In fact, when we have data that needs to be processed in a model column, we usually have `serialize` / `parse` functions for it (for example serializeOcrDetails, serializeUserData), so please add similar functions to handle your extracted_resource_ids. This is also an opportunity to validate the data and ensure that for example a third-party tool or plugin didn't insert a list with spaces or other invalid data.

File: packages/lib/services/ResourceService.ts (hunk 1 above)
Comment (laurent22) [reply packages/lib/services/ResourceService.ts]: Regarding this issue, please could you add a comment why it makes sense to create this fake note body and parse it using setAssociatedResources?

File: packages/lib/services/ResourceService.test.ts (hunk 3 above)
Comment (keshav0479) [reply packages/lib/services/ResourceService.test.ts]: i was following the `isNoteLockEnabled()` guard added in #15673, but kept this indexer handling outside it because locked notes can still exist or arrive through sync when the flag is off, scanning their ciphertext would clear the resource associations and could make attachments look orphaned.

[this proposal comment](https://docs.google.com/document/d/1YU3lY1ZmtFvMTKZDUyRmX4cB7ly8pzO8ECu2xacfqng/edit?disco=AAAB9KT_cS8) also calls out keeping this protection active with the flag off. should this stay as an exception, or would it be better behind the guard?

Comment (laurent22) [reply packages/lib/services/ResourceService.test.ts]: Please setup the flag in a way that makes sense, you know the project better than me

The fact that you're currently using the env variable for this is a problem and, as in this example, it's starting to leak in the code. We generally don't want to dynamically switch the env even in tests

Comment (keshav0479) [reply packages/lib/services/ResourceService.test.ts]: got it, i'll replace the env check with a proper flag and update the tests to use that instead. that way the ResourceService safety handling can still run when the flag is off

Comment (keshav0479) [reply packages/lib/services/ResourceService.test.ts]: replaced the env check with a proper feature flag and updated tests to use it directly.

File: packages/lib/services/ResourceService.ts (hunk 1 above)
Comment (keshav0479) [reply packages/lib/services/ResourceService.ts]: added a short comment explaining that generated note links reuse existing resource validation

File: packages/lib/services/ResourceService.ts (hunk 4 above)
Comment (keshav0479) [reply packages/lib/services/ResourceService.ts]: done, added serialize/unserialize helpers that handle empty values and filter invalid or spaced IDs

File: packages/lib/models/settings/builtInMetadata.ts [hunk 5]
```diff
@@ -2233,6 +2233,13 @@ const builtInMetadata = (Setting: typeof SettingType) => {
 			appTypes: [AppType.Mobile],
 		},
 
+		'featureFlag.noteLock': {
+			value: false,
+			type: SettingItemType.Bool,
+			public: false,
```
Comment (personalizedrefrigerator) [inline packages/lib/models/settings/builtInMetadata.ts:2239]: **Note**: The `public: false` seems to be ignored on mobile. I've opened an issue to track this here: https://github.com/laurent22/joplin/issues/15729. (`featureFlag.noteLock` may be shown in the mobile settings screen, despite the lack of a description/label and the `public:false` modifier.)

Comment (keshav0479) [reply packages/lib/models/settings/builtInMetadata.ts:2239]: thanks for catching this, i opened #15730 to fix the mobile filtering, so private flags like this one won’t show in mobile settings

File: packages/lib/models/Note.ts [hunk 6]
```diff
@@ -175,6 +176,15 @@ export default class Note extends BaseItem {
...
 	}
 
+	public static serializeExtractedResourceIds(resourceIds: string[]) {
+		return unique(resourceIds.map(id => id.trim()).filter(id => !!isItemId(id))).join(',');
+	}
+
+	public static unserializeExtractedResourceIds(serializedIds: string) {
+		if (!serializedIds) return [];
+		return unique(serializedIds.split(',').map(id => id.trim()).filter(id => !!isItemId(id)));
+	}
```
Comment (laurent22) [inline packages/lib/models/Note.ts:186]: please add tests for this using `test.each`. Include things like single space, empty string, double comma, etc.

Comment (keshav0479) [reply packages/lib/models/Note.ts:186]: done, split this into focused `test.each` cases, including the edge cases you mentioned
